src/wp-includes/theme.php

```diff
@@ -321,24 +330,33 @@ function get_template() {
  * Retrieves template directory path for the active theme.
  *
  * @since 1.5.0
+ * @since 6.4.0 Memoizes filter execution so that it only runs once for the current theme.
+ *
+ * @global string $wp_template_path Current theme template directory path.
  *
  * @return string Path to active theme's template directory.
  */
 function get_template_directory() {
-	$template     = get_template();
-	$theme_root   = get_theme_root( $template );
-	$template_dir = "$theme_root/$template";
+	global $wp_template_path;
 
-	/**
-	 * Filters the active theme directory path.
-	 *
-	 * @since 1.5.0
-	 *
-	 * @param string $template_dir The path of the active theme directory.
-	 * @param string $template     Directory name of the active theme.
-	 * @param string $theme_root   Absolute path to the themes directory.
-	 */
-	return apply_filters( 'template_directory', $template_dir, $template, $theme_root );
+	if ( null === $wp_template_path ) {
```

Comment:
```suggestion
	if ( null === $wp_template_path || has_filter( 'template_directory' ) ) {
```

Reply:
See my reply in https://github.com/WordPress/wordpress-develop/pull/5160#discussion_r1323519557.